### PR TITLE
Optimistic Confirmation Proposal

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
No proposal for optimistic confirmation

#### Summary of Changes
Add such a proposal

Fixes #

### CODE REVIEW

File: docs/src/proposals/optimistic_confirmation.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+and for all `s` in `S`, `s + lockout(s) < X'`
+* `X > X'` implies `X > S'.last` and `S.last > S'.last` 
+and for all `s` in `S'`, `s + lockout(s) < X`
+
+Otherwise the validator is slashed. 
+
+`Range(vote)` - Given a vote `v = vote(X, S)`, define `Range(v)` to be the range
+ of slots `[X, S.last]`.
+
+`SP(old_vote, new_vote)` - This is the "Switching Proof" for `old_vote`, the
```
Comment (carllin) [inline docs/src/proposals/optimistic_confirmation.md]: @aeyakovenko in this proposal it's assumed every switch to a different fork will necessarily generate a "switching proof". 

!hat's not covered in this proposal is 

1) Will generating these proofs on every switch be inefficient. 
2) When will such a switching proof be verified/a challenged issued? Every time?

I think we need to think very carefully about the above...

Comment (aeyakovenko) [reply docs/src/proposals/optimistic_confirmation.md]: The proof is just the image.  I don’t think it needs to be uploaded every time.

File: docs/src/proposals/optimistic_confirmation.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+(Recall the vote should contain the `latest` switch).
+
+To enforce this, we define the "Optimistic Slashing" slashing conditions. Given
+any two distinct votes `vote(X, S)`and `vote(X', S')` by the same validator,
+the votes must satisfy (namely the last two rules imply the ranges cannot
+overlap):
+
+* `X <= S.last`, `X' <= S'.last`
+* All `s` in `S.last` are ancestors/descendants of one another, 
+all `s'` in `S'.last` areancsestors/desccendants of one another,
```
Comment (aeyakovenko) [inline docs/src/proposals/optimistic_confirmation.md]: `areaancestors` needs a space

Comment (aeyakovenko) [reply docs/src/proposals/optimistic_confirmation.md]: X must be an ancestor of a slot in S?

File: docs/src/proposals/optimistic_confirmation.md (hunk 1 above)
Comment (aeyakovenko) [reply docs/src/proposals/optimistic_confirmation.md]: We only need the challenge if someone can present a proof of a block with a supermajority vote that is not an ancestor of the current fork.

Comment (aeyakovenko) [reply docs/src/proposals/optimistic_confirmation.md]: The way i imagine this is implemented, is someone uploads this thing to the cluster, and it takes up to the end of the following epoch to submit the challenges.

File: docs/src/proposals/optimistic_confirmation.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+`Range(vote)` - Given a vote `v = vote(X, S)`, define `Range(v)` to be the range
+ of slots `[X, S.last]`.
+
+`SP(old_vote, new_vote)` - This is the "Switching Proof" for `old_vote`, the
+validator's latest vote. Such a proof is necessary for a validator to vote 
+on any slot `new_vote.last` that is not a descendant of `old_vote.last`. Note 
+switching must still respect lockouts.
+
+A switching proof shows that `> 1/3` of the network is locked out at slot
```
Comment (aeyakovenko) [inline docs/src/proposals/optimistic_confirmation.md]: `1/3 + Z`, where Z is what we lose for liveness.

**Comment by aeyakovenko** [comment]
I need some drawing for those proofs :)

File: docs/src/proposals/optimistic_confirmation.md (hunk 2 above)
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: Of every slot in S, it's the slot at which this validator last "switched" to this fork

**Comment by carllin** [comment]
@aeyakovenko Hehe totally fair, this is pretty much a brain dump, need to organize some pictures.

File: docs/src/proposals/optimistic_confirmation.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+of votes must have been: `X ... S' ... S`. This means after the second vote the
+validator must have switched back to the original fork at some `X' > X`. This 
+means the vote for `S` should have used `X'` as the "reference" point, so the 
+vote should have been of the form `vote(X', S)`, not `vote(X, S)`
+(Recall the vote should contain the `latest` switch).
+
+To enforce this, we define the "Optimistic Slashing" slashing conditions. Given
+any two distinct votes `vote(X, S)`and `vote(X', S')` by the same validator,
+the votes must satisfy (namely the last two rules imply the ranges cannot
+overlap):
```
Comment (rwalker-com) [inline docs/src/proposals/optimistic_confirmation.md]: I think if you moved this parenthetical phrase down, it'd be easier to read

File: docs/src/proposals/optimistic_confirmation.md (hunk 2 above)
Comment (rwalker-com) [inline docs/src/proposals/optimistic_confirmation.md]: Do you mean "all `s` in `S`..." for these two lines?  `S.last` is a single slot, right?

File: docs/src/proposals/optimistic_confirmation.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+validator's latest vote. Such a proof is necessary for a validator to vote 
+on any slot `new_vote.last` that is not a descendant of `old_vote.last`. Note 
+switching must still respect lockouts.
+
+A switching proof shows that `> 1/3` of the network is locked out at slot
+`old_vote.last`.
+
+The proof is a list of elements `(validator_id, validator_vote(X, S))`, where:
+
+1. The sum of the stakes of all the validator id's `> 1/3`
```
Comment (rwalker-com) [inline docs/src/proposals/optimistic_confirmation.md]: for a switching proof to be valid on a given fork, the stakes must match the fork being switched away from, right?

File: docs/src/proposals/optimistic_confirmation.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+Switching forks without a valid switching proof is slashable.
+
+## Definitions:
+
+Optimistic Confirmation - A block `B` is then said to have achieved
+"optimistic confirmation" if `>2/3` of stake have voted with votes `v`
+where `Range(v)` for each such `v` includes `B.slot`.
+
+Finalized - A block `B` is said to be finalized if at least one 
+correct validator has rooted `B` or a descedant of `B`.
```
Comment (rwalker-com) [inline docs/src/proposals/optimistic_confirmation.md]: descendant

Comment (rwalker-com) [reply docs/src/proposals/optimistic_confirmation.md]: what is a "correct validator"?

Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: It's a validator that is following the protocol as written, not acting in a byzantine way.

File: docs/src/proposals/optimistic_confirmation.md (hunk 2 above)
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: ah yeah, thanks! 😄

File: docs/src/proposals/optimistic_confirmation.md (hunk 4 above)
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: Done, thanks!

File: docs/src/proposals/optimistic_confirmation.md (hunk 5 above)
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: Yes, I think we should be using the EpochStakes for the epoch that slot is in.

File: docs/src/proposals/optimistic_confirmation.md (hunk 1 above)
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: > We only need the challenge if someone can present a proof of a block with a supermajority vote that is not an ancestor of the current fork.

Hmmm not sure  understand this as phrased. So only when you are switching to a supermajority fork you upload an image? Might need an example here to illustrate this...

Comment (aeyakovenko) [reply docs/src/proposals/optimistic_confirmation.md]: No. I mean most of the time no one is cheating, or not enough cheat.  The only time proofs need to be presented is if you can show there was an older block that got 2/3+ votes that’s not an ancestor of the current fork.

File: docs/src/proposals/optimistic_confirmation.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
+# Optimistic Confirmation
+
+## Primitives
+
+`vote(X, S)` - Votes will be augmented with a "reference" slot, `X` 
+which is the **last** ancestor at which this validator voted on this fork
+with a proof of switching. Thus as long as a validator keeps voting on the 
+same fork, `X` will remain unchanged.All votes will then be of the form
+`vote(X, S)`, where `S` is the sorted list of slots `(s, s.lockout)`
```
Comment (aeyakovenko) [inline docs/src/proposals/optimistic_confirmation.md]: Is S the entire Tower stack?

File: docs/src/proposals/optimistic_confirmation.md [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+vote should have been of the form `vote(X', S)`, not `vote(X, S)`
+(Recall the vote should contain the `latest` switch).
+
+To enforce this, we define the "Optimistic Slashing" slashing conditions. Given
+any two distinct votes `vote(X, S)`and `vote(X', S')` by the same validator,
+the votes must satisfy:
+
+* `X <= S.last`, `X' <= S'.last`
+* All `s` in `S` are ancestors/descendants of one another, 
+all `s'` in `S'` areancsestors/descendants of one another,
```
Comment (aeyakovenko) [inline docs/src/proposals/optimistic_confirmation.md]: `areancestors`

File: docs/src/proposals/optimistic_confirmation.md (hunk 7 above)
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: @aeyakovenko, I think that will depend on how we define the contents of the proof of switching, specifically how we intend prove the amount of "lockout" a node has committed to a specific slot `s`.

Right now the section about the contents of the `proof of switching` here: https://github.com/solana-labs/solana/pull/9157/files#diff-af0aef1677e9797c44811a991da7fa2eR54-R64

 is pretty hand-wavy in that it assumes there's some to show in a `proof of switching` that validator `X` is locked out on slot `s` at slot `s'`, i.e. show that `s + s.lockout > s'`. 

This means either introducing some sort of new `lockout` field to our current votes (how do you verify a node is not lying about their lockout), or the proof of switching will need to include a series of votes that demonstrate lockout was built up.

File: docs/src/proposals/optimistic_confirmation.md [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+From above, `S.last >= B >= X` so for all such "switching votes", `X_v > B`.
+
+Ordering all these "switching votes" in time, let `V` to be the validator
+that first submitted such a "swtching vote" `Vote(X', S')`, where from the
+properties derived from above, `X' > B`
+
+Now let `Vote(X, S)` be the vote in `Delinquent Optimistic Votes` made by
+validator `V`. 
+
+Given `Vote(X, S)` and the the "Optimistic Slashing" rules, because
```
Comment (jstarry) [inline docs/src/proposals/optimistic_confirmation.md:346]: nit: the the

File: docs/src/proposals/optimistic_confirmation.md [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+`X' > B > X`, then `X' > S.last`.
```
Comment (jstarry) [inline docs/src/proposals/optimistic_confirmation.md:347]: should this be: `B >= X`?

File: docs/src/proposals/optimistic_confirmation.md [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+
+In order to perform such a "switching vote" to `X'`, a switching proof
+`SP(Vote(X, S), Vote(X', S')` must show `> 1/3` of stake being locked
```
Comment (jstarry) [inline docs/src/proposals/optimistic_confirmation.md:350]: nit, missing paren: `SP(Vote(X, S), Vote(X', S'))`

File: docs/src/proposals/optimistic_confirmation.md [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+out at this validator's latest vote, `S.last`. Combine this `>1/3` with the
+fact that the set of validators in `Delinquent` consists of `> 2/3` of the
```
Comment (jstarry) [inline docs/src/proposals/optimistic_confirmation.md:352]: I thought `Delinquent` represented the 1/3 of stake that both rooted B' and was included in the optimistic confirmation of B?

Should this be: "that the set of validators in `Delinquent`" -> "the set of validators who rooted `B'`"?

File: docs/src/proposals/optimistic_confirmation.md [hunk 13]
```diff
@@ -0,0 +1,380 @@
+# Optimistic Confirmation
+
+## Primitives
+
+`vote(X, S)` - Votes will be augmented with a "reference" slot, `X` 
+which is the **last** ancestor at which this validator voted on this fork
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md:6]: much better, thanks!

[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md:6]: The wording is much clearer! Note it's the "latest" switch, not the "earliest", so I've updated to:
> Votes will be augmented with a "reference" slot, `X`, which is the **latest** ancestor of this fork that this validator voted on with a proof of switching.

File: docs/src/proposals/optimistic_confirmation.md [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+Given a vote `vote(X, S)`, let `S.last == vote.last` be the last slot in `S`.
+
+Now we define some "Optimistic Slashing" slashing conditions. The intuition
+for these is described below:
+
+* `Intuition`: If a validator submits `vote(X, S)`, the same validator
+should not have voted on a different fork that "overlaps" this fork. 
+More concretely, this validator should not have cast another vote
+`vote(X', S')` where the range `[X, S.last]` overlaps the range
+`[X', S'.last]`, as shown below:
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md:21]: very true, fixed

File: docs/src/proposals/optimistic_confirmation.md [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+                        |
+                    +---+---+
+                    |       |
+             S.last |       |
+                    |       |
+                    +-------+
+                    
+(Example of slashable votes vote(X', S') and vote(X, S))
+
+In the diagram above, note that the vote for `S.last` must have come after the
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md:70]: True, changed to "must have been **sent** after" as this refers to the order in which the same validator sent two different votes

File: docs/src/proposals/optimistic_confirmation.md (hunk 12 above)
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: @jstarry you're totally right this part was completely unclear. I've reworked it in my latest commit, what do you think?

File: docs/src/proposals/optimistic_confirmation.md (hunk 10 above)
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: yup! fixed

File: docs/src/proposals/optimistic_confirmation.md [hunk 16]
```diff
@@ -0,0 +1,397 @@
+# Optimistic Confirmation
+
+## Primitives
+
+`vote(X, S)` - Votes will be augmented with a "reference" slot, `X`
+which is the **latest** ancestor of this fork that this validator voted on
+with a proof of switching. Thus as long as a validator keeps voting on the
+same fork, `X` will remain unchanged.All votes will then be of the form
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: Clarified...hopefully 😄

File: docs/src/proposals/optimistic_confirmation.md (hunk 12 above)
Comment (jstarry) [reply docs/src/proposals/optimistic_confirmation.md]: Makes more sense now! Left some comments to make sure I'm understanding things correctly

File: docs/src/proposals/optimistic_confirmation.md [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+Now consider if `B'` < `X`:
+
+`Case B' < X`: We wll show this is a violation of lockouts.
+From above, we know `B'` is not a parent of `X`. Then because `B'` was rooted,
+and `B'` is not a parent of `X`, then the validator should not have been able
+to vote on the higher slot `X` that does not descend from `B'`.
+
+### Proof of Safety:
+We now aim to show at least one of the validators in the
+`Optimistic Validators` set violated a slashing rule.
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md:327]: that's exactly right!

**Comment by carllin** [comment]
@jstarry close! The rough idea is that the switching proof shows there is sufficient stake on some other forks (not the one you're on) to prevent the fork you're currently voting on from getting > 2/3 (aka there is a non-zero chance some other fork could get confirmed)

File: docs/src/proposals/optimistic_confirmation.md [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+`Claim:` Given a vote `Vote(X, S)` made by a validator `V` in the
+`Optimistic Validators` set, and `S` contains a vote for a slot `s`
+for which:
+
+* `s + s.lockout > B`,
+* `s` is not an ancestor or descendant of `B`,
+
+then `X > B`.
+
+`
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md:175]: Fixed!

[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/proposals/optimistic_confirmation.md]: Fixed!

**Comment by jstarry** [comment]
@carllin I'm curious why you chose to use switching proofs rather than the opposite.. slashing proofs for delinquent switching

**Comment by jstarry** [comment]
Also, do we need reference slots? Can't other nodes detect when a validator votes on a new fork?

**Pull request merged**